core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1739,6 +1758,37 @@ class ReplicaManager(val config: KafkaConfig,
     partitionsToMakeFollower
   }
 
+  private def updateTopicIdForFollowers(controllerId: Int,
+                                        controllerEpoch: Int,
+                                        partitionStates: Set[Partition],
+                                        correlationId: Int,
+                                        topicIds: String => Option[Uuid]): Unit = {
+    val traceLoggingEnabled = stateChangeLogger.isTraceEnabled
+
+    try {
+      if (isShuttingDown.get()) {
+        if (traceLoggingEnabled) {
+          partitionStates.foreach { partition =>
+            stateChangeLogger.trace(s"Skipped the update topic ID step of the become-follower state " +
+              s"change with correlation id $correlationId from controller $controllerId epoch $controllerEpoch for " +
+              s"partition ${partition.topicPartition} since it is shutting down")
+          }
+        }
+      } else {
+        val partitionsToUpdateFollowerWithLeader = partitionStates.map { partition =>
+          partition.topicPartition -> partition.leaderReplicaIdOpt.getOrElse(-1)
```

Comment:
nit: Should we simply ignore a partition without leader instead of using `-1`? This case should never happen, I think. We could do something like this:

```
val partitionsToUpdateFollowerWithLeader = mutable.Map.empty[TopicPartition, Int]
partitionStates.foreach { partition =>
  partition.leaderReplicaIdOpt.foreach { leader =>
    if (metadataCache.hasAliveBroker(leader)) {
      partitionsToUpdateFollowerWithLeader += partition.topicPartition -> leader
    }
  }
}

```

Reply:
I was wondering this too, but didn't know the best way. This makes sense to me.